File: app/src/main/java/com/qamar4p/loladapterexample/ScrollingActivity.java
package com.qamar4p.loladapterexample;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.qamar4p.lol_adapter.ItemViewClickListener;
import com.qamar4p.lol_adapter.LolAdapter;
import com.qamar4p.lol_adapter.LolTypeAdapter;
import com.qamar4p.lol_adapter.LolViewHold;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.OnClick;

/**
 * @author Qamar4P
 */
public class ScrollingActivity extends BaseActivity{

    public static final String AUTHOR_CELL = "+923088006866";
    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;
    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_scrolling);
        setSupportActionBar(toolbar);

        setupUi();
        loadData();
    }

    @OnClick(R.id.fab)
    public void tapOnFab(View view){
        startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("smsto:"+ AUTHOR_CELL)));
    }

    private void setupUi() {
        recyclerView.setAdapter(adapter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_scrolling, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_single_view_type) {
            recyclerView.setAdapter(adapter);
            return true;
        }
        if (id == R.id.action_multi_view_type) {
            recyclerView.setAdapter(typeAdapter);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void loadData() {
        ArrayList<ItemModel> data = new ArrayList<>();
        data.add(new ItemModel());
        data.add(new ItemModel());
        data.add(new ItemModel());
        data.add(new ItemModel());
        data.add(new ItemModel());
        data.add(new ItemModel());
        data.add(new ItemModel());
        data.add(new ItemModel());
        data.add(new ItemModel());

        adapter.items.addAll(data);
        typeAdapter.items.addAll(data);
    }

    @NonNull
    private ItemViewClickListener<ItemModel> itemViewClickListener() {
        return (v, item, position) -> {
            switch (v.getId()) {
                case R.id.buttonCall:
                    startActivity(new Intent(Intent.ACTION_DIAL, Uri.parse("tel:"+ AUTHOR_CELL)));
                    break;
                default:
                    break;
            }
        };
    }

    LolAdapter<ItemModel,ItemViewHold> adapter = new LolAdapter<>(itemViewClickListener(), ItemViewHold::new);

    /**
     * recycle view adapter for multiple view types
     */
    LolAdapter<ItemModel,LolViewHold<ItemModel>>
            typeAdapter = new LolTypeAdapter<>(itemViewClickListener()
            ,position -> position == 0 || position == 4 ? 0 : 1,
            ItemHeaderView::new,
            ItemViewHold::new);

    class ItemViewHold extends LolViewHold<ItemModel> {
        @BindView(R.id.textTitle)
        TextView textTitle;
        @BindView(R.id.textDesc)
        TextView textDesc;

        ItemViewHold(ViewGroup parent) {
            super(parent,R.layout.list_item);
        }
        @Override
        public void bind() {
            textTitle.setText(data.fullName+" "+getAdapterPosition());
            textDesc.setText(data.location);
        }

        @OnClick({R.id.buttonCall})
        public void tapOnView(View v){
            viewClickListener.onViewClicked(v,data,getAdapterPosition());
        }
    }

    class ItemHeaderView extends LolViewHold<ItemModel> {
        @BindView(R.id.textTitle)
        TextView textTitle;

        ItemHeaderView(ViewGroup parent) {
            super(parent,R.layout.list_item2);
        }

        @Override
        public void bind() {
            textTitle.setText(data.fullName+" "+getAdapterPosition());
        }
    }
}

File: app/src/main/java/com/qamar4p/loladapterexample/ScrollingActivity.kt
package com.qamar4p.loladapterexample

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import java.util.ArrayList

import com.qamar4p.lol_adapter.ItemViewClickListener
import com.qamar4p.lol_adapter.LolAdapter
import com.qamar4p.lol_adapter.LolTypeAdapter
import com.qamar4p.lol_adapter.LolViewHold

import kotlinx.android.synthetic.main.activity_scrolling.*
import kotlinx.android.synthetic.main.content_scrolling.*
import kotlinx.android.synthetic.main.list_item.view.*

/**
 * @author Qamar4P
 */
class ScrollingActivity : BaseActivity() {

    private var adapter = LolAdapter<ItemModel, ItemViewHold>(itemViewClickListener(),{ItemViewHold(it)})

    /**
     * recycle view adapter for multiple view types
     */
    private var typeAdapter: LolAdapter<ItemModel, LolViewHold<ItemModel>>
            = LolTypeAdapter(itemViewClickListener(), {
        if (it == 0 || it == 4) 0 else 1
    },{ItemHeaderView(it) },
      {ItemViewHold(it)})

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_scrolling)
        setSupportActionBar(toolbar)

        loadData()
        setupUi()
    }

    private fun setupUi() {
        recyclerView!!.adapter = adapter
        fab.setOnClickListener({
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("smsto:" + AUTHOR_CELL)))
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_scrolling, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId


        if (id == R.id.action_single_view_type) {
            recyclerView!!.adapter = adapter
            return true
        }
        if (id == R.id.action_multi_view_type) {
            recyclerView!!.adapter = typeAdapter
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun loadData() {
        val data = arrayListOf<ItemModel>()
        data.add(ItemModel())
        data.add(ItemModel())
        data.add(ItemModel())
        data.add(ItemModel())
        data.add(ItemModel())
        data.add(ItemModel())
        data.add(ItemModel())
        data.add(ItemModel())
        data.add(ItemModel())

        adapter.items += data
        typeAdapter.items += data
    }

    private fun itemViewClickListener(): ItemViewClickListener<ItemModel> {
        return object:ItemViewClickListener<ItemModel>{
            override fun onViewClicked(v: View, item: ItemModel?, position: Int) {
                when (v.getId()) {
                    R.id.buttonCall -> startActivity(Intent(Intent.ACTION_DIAL, Uri.parse("tel:" + AUTHOR_CELL)))
                    else -> {
                    }
                }
            }
        }
    }

    companion object {
        val AUTHOR_CELL = "+923088006866"
    }
}

class ItemViewHold(parent: ViewGroup) : LolViewHold<ItemModel>(parent, R.layout.list_item) {

    init {
        itemView.buttonCall.setOnClickListener { viewClickListener?.onViewClicked(it, data, adapterPosition) }
    }

    override fun bind() {
        itemView.textTitle!!.text = data?.fullName + " " + adapterPosition
        itemView.textDesc!!.text = data?.location
    }
}

class ItemHeaderView(parent: ViewGroup) : LolViewHold<ItemModel>(parent, R.layout.list_item2) {
    override fun bind() {
        itemView.textTitle!!.text = data?.fullName + " " + adapterPosition
    }
}
